스케줄 도메인 홈, 워크스페이스 페이지 조회 API 구현

jkde7721 opened the pull request:
### 기능적인 부분에 대해서는 일단 구현 완료, 앞으로 할 것
- 아키텍처 관련해서 문서로 정리 후 profile, workspace, schedule 도메인에 적용할 예정
- Beacon 구현 
- 예외 처리 
- 프론트와 연결하면서 API 추가 구현 필요

jkde7721 committed:
feat: Staff 홈 달력 정보 조회 API 구현

jkde7721 committed:
feat: Staff 해당 워크스페이스 홈 달력 정보 조회 API 구현

jkde7721 committed:
feat: Boss 홈 달력 정보 조회 API 구현

jkde7721 committed:
feat: Boss 해당 워크스페이스 홈 달력 정보 조회 API 구현

jkde7721 committed:
fix: API path 오타 수정

jkde7721 committed:
fix: 달력 정보 조회 API에서 dateList 조회 관련 문제 해결

- schedule이 존재하는 date만 조회되었음. 즉 12/25에만 스케줄 존재한다면
  12/1 ~ 12/31까지 조회되는 것이 아닌 12/25만 dateList에 포함되던 문제
- dateList 날짜순으로 정렬되지 않던 문제

jkde7721 committed:
refactor: dto 구조 및 이름 변경

jkde7721 committed:
refactor: dto 이름, 필드 순서 및 mapper 메소드 이름, 정의 순서 변경

jkde7721 committed:
!BREAKING CHANGE: Date 관련 response dto에 요일(dayOfWeek) 필드 추가

jkde7721 committed:
refactor: 외부에서 호출되지 않는 repository 메소드 제거 후 복잡한 쿼리 impl로 옮김

jkde7721 committed:
refactor: schedule 엔티티 양방향 연관관계 제거

schedule ↔️ substituteReq, timecardReq 간 양방향 연관관계 제거

jkde7721 committed:
add: ScheduleStatus 설명 주석 추가

jkde7721 committed:
refactor: ScheduleReportService가 의존하던 WorkspaceMemberRepository 제거 후, StaffProfileService 의존

jkde7721 committed:
!BREAKING CHANGE: workspace 도메인 API → schedule 도메인 API로 변경

해당 워크스페이스 내 알바 리스트 조회 (for 스케줄 추가): /boss/workspaces/{workspaceId}/schedule/staffs
➡️ 해당 날짜에 근무 가능한 해당 워크스페이스 내 알바 리스트 조회 (for 스케줄 추가): /boss/schedules/workspaces/{workspaceId}/staffs/possible/for

jkde7721 committed:
refactor: ScheduleDateTime에 대한 반복 제거

ScheduleDateTime에 검증 어노테이션 적용 후, ScheduleCreateRequest에서
기존 scheduleStartDateTime, scheduleEndDateTime 필드를 ScheduleDateTime으로 대체

jkde7721 committed:
!BREAKING CHANGE: 해당 날짜에 근무 가능한 해당 워크스페이스 내 알바 리스트 조회 API 변경

날짜 정보 기존에는 요청 body에 담아서 전달
➡️ 쿼리 스트링으로 전달로 변경 (날짜 파싱 위해 @DateTimeFormat 어노테이션 추가)

jkde7721 committed:
fix: schedule 생성 dto 변경에 따른 mapper 변경

jkde7721 commented:
**일단 현재 workspace 도메인에 정의되어있던 API 1개가 schedule 도메인 쪽으로 옮겨진 상태** 
(스케줄 추가 전) 해당 날짜에 근무 가능한 해당 워크스페이스 내 알바 리스트 조회 API 

이전에는 **워크스페이스** 내 알바 리스트 조회라고 생각해서 workspace 도메인에 포함하였는데 **해당 날짜**라는 부분이 결국 스케줄 날짜가 될 것이고 
이 부분이 스케줄 생성 API에 적용되는 스케줄 날짜에 대한 validation(스케줄 날짜는 오늘 날짜로부터 +1 ~ +90, 스케줄 시간은 10분 단위, 스케줄 시작 시간 < 스케줄 종료 시간 등의 조건)과 중복되어 해당 API schedule 도메인으로 옮김

뿐만 아니라 현재 workspace 도메인에 정의된 API 중 실제 사용되는 도메인은 message, substituteReq 쪽인데 workspace 도메인 내에 정의되어 있는 경우가 있음 ➡️ 나중에 옮길 예정

**ex.** `StaffWorkspaceController` 내에 선언된 **해당 스케줄 가능한 알바 리스트 조회** API는 대타근무 요청 도메인에서 사용 ➡️ **해당 스케줄 가능한 알바 리스트 조회 + 기존에 해당 스케줄에 대해 대타근무 요청을 받고 요청 상태가 진행 중인 알바는 제외**라는 식으로 요구사항이 구체화되면서 대타근무 요청 도메인과 관련이 짙어질 수 있음 
**ex.** `BossWorkspaceController` 내에 선언된 **해당 워크스페이스 내 알바 리스트 조회 (for 쪽지함)** ➡️ 현재는 Staff 관련 데이터만을 응답으로 주지만 추후에 Message 관련 데이터 또한 응답에 포함될 수 있음 

➡️ 이러한 이유로 해당 API 2개 도메인 위치 바꿀 예정

jiyeoon00 commented on src/main/java/com/project/coalba/domain/schedule/service/StaffScheduleService.java:
```diff
@@ -23,20 +28,54 @@ public class StaffScheduleService {
+        if (scheduleList == null) return new HomeDateServiceDto(date, false, false, false);
+        if (date.isAfter(LocalDate.now())) return new HomeDateServiceDto(date, true, true, false);
+        if (isAllSuccess(scheduleList)) return new HomeDateServiceDto(date, true, false, true);
+        return new HomeDateServiceDto(date, true, false, false);
```
여기서 status를 알아내는데 필요한 매개변수가 true, false등 3개 조건으로 나와있고 dto매핑할때 이 조건들을 가지고 상태값을 가져오는거 같은데 , 각 매개변수의 true, false에 대한 의미를 하나하나 봐야해서 가독성이 떨어지는거 같아! 
HomeDateServiceDto가 3개의 Boolean변수를 가지는 것 보다 TotalScheduleStatus Enum값(COMPLETE, INCOMPLETE, BEFORE, NONE)을 가지고 있어서 여기서(=getHomeDate메소드에서) 상태값을 아예 지정한 후 리턴해주던가 
아니면 false, true값들을 넘기는게 아니라 List<Schedule>을 ScheduleMapper의 getTotalScheduleStatus메소드의 매개변수로 받아서 여기서 바로 상태값을 리턴해주는건 어때?!

jkde7721 replied:
`TotalScheduleStatus` enum이 응답용 데이터라고 생각해서 schedule/dto/response/enums 패키지(web 계층) 내부에 정의해놔서 ServiceDto(application 계층)에서 해당 enum 의존하지 않도록 했는데, 생각해보니 기존 `ScheduleStatus` 값을 통해 `TotalScheduleStatus`를 도출해내는 로직이 있으니깐 이 또한 비즈니스 로직과 관련이 있다고 판단해서 schedule/entity/enums 내부로 패키지 이동했어. 그리고 첫번째 솔루션대로 `HomeDateServiceDto`, `BossWorkspaceDateServiceDto`에서 `TotalScheduleStatus` 값 가지고 있도록 수정했어!

jkde7721 commented:
### ServiceDto 사용 관련 
- **controller ➡️ service**
  - `@PathVariable`, `@RequestParam`로 넘어온 데이터를 service로 전달하는 경우를 제외하고 service로 데이터를 넘기는 대부분은 Create, Update를 위해서임
  - 이전에는 controller에서 클라이언트로부터 받은 Dto를 mapper 통해 entity로 변환하여 service로 전달함, 이때 변환된 entity는 불완전함(몇몇 객체 필드가 set되지 않은 상태)
  - 차라리 중간 단계인 ServiceDto를 두고 service 계층에서 해당 ServiceDto를 entity로 변환하기로 함 
  - ex. profile, workspace, schedule 도메인에서의 Create, Update API
 - **service ➡️ controller**
   - 대부분 이 방향의 응답은 service에서 entity를 그대로 반환하고 mapper를 통해 Dto로 변환하는 방식을 채택
   - Dto가 복잡한 경우 **entity 그 자체 + 별도의 부가 정보** 또는 A entity + B entity와 같은 **여러 entity 조합**을 controller에게 넘겨줘야 함 
   - 이 경우 별도의 ServiceDto를 두고 이를 mapper에서 Dto로 변환함 
   - ex. `StaffScheduleController` 내의 **해당 워크스페이스 홈 해당 날짜 스케줄 조회 API**에서 해당 날짜의 스케줄 entity 뿐만 아니라 해당 스케줄이 자신(현재 로그인한 유저)의 스케줄인지 여부를 나타내는 `isMySchedule` 필드를 포함하는 `ScheduleServiceDto` 사용 

### DTO 클래스 구조 관련 
```json
"staffList": [
  {
    "staffId": 1,
    "name": "abc",
    "imageUrl": "http://"
  },
  {
    "staffId": 2,
    "name": "def",
    "imageUrl": "http://"
  }
]
```
```java
public class StaffListResponse {
  private List<StaffResponse> staffList;

  public static class StaffResponse {
    private Long staffId;
    private String name;
    private String imageUrl; 
  }
}
```
- 현재 응답 Dto 내부에 List 타입 필드에 사용되는 클래스는 따로 외부 클래스로 정의하지 않고 해당 응답 Dto 클래스(`StaffListResponse`)의 내부 클래스(`StaffResponse`)로 정의함 
- 다른 응답 Dto에서도 반복적으로 사용되고 변경의 가능성이 적은 응답 Dto는 외부 클래스로 정의 ex. `DateResponse`, `HomeDateResponse` 
- `BossWorkspaceScheduleListResponse`, `StaffWorkspaceScheduleListResponse` 이 2개를 비교해보면 중복되는 데이터가 많고, 특히 내부 클래스인 `WorkerResponse`는 외부 클래스로 빼도 되지 않을까 생각했지만 추후에 Boss 쪽에만 혹은 Staff 쪽에만 `WorkerResponse`에 대한 변경이 발생할 가능성이 높기 때문에 따로 분리하지 않음 
➡️ 변경 가능성이 높은 Dto를 외부 클래스로 분리하고 여러 다른 응답 Dto가 해당 Dto(변경 가능성이 높은 Dto)를 의존하게 되면 변경에 따른 영향력이 크다고 생각함

- **schedule/dto/response** 패키지 내 클래스 구조가 조금 복잡한데 <br>
<img src="https://user-images.githubusercontent.com/65665065/210757659-d0d30774-cf1a-48f6-a3dd-59479cfe4dad.png" width="180" height="365"/>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<img src="https://user-images.githubusercontent.com/65665065/210757674-1059812e-cfae-43c0-bf11-af281b0d9ed9.png" width="180" height="365"/>
- 위 그림처럼 전체 데이터는 **Page**(**Date** + **ScheduleList**)로 구성되어 있다고 생각하면 돼
- 각 날짜를 선택할 때마다 새로운 ScheduleList 데이터로 업데이트 되는 거고!

### Service 구조 관련 
- 단순히 scheduleId로 해당 Schedule entity를 조회하는 로직을 다른 도메인에서도 의존하여 사용하는 경우가 많은데 이를 `BossScheduleService` 혹은 `StaffScheduleService` 쪽에 넣을지 모르겠어서 따로 `ScheduleService` 클래스로 분리 
- `StaffScheduleService`에서도 scheduleId로 해당 Schedule entity를 조회하는 로직을 사용하기 때문에 해당 메소드를 중복 정의하긴 했는데, 어떻게 할지 고민임... (다른 도메인에서도 이러한 경우 존재)

### Mapper 코드 리뷰 관련 
응답 JSON 데이터 자체가 복잡하고 계속 overloading 하다 보니 너무 복잡해져서 이 부분은 시간 없으면 리뷰 건너뛰어도 될 듯!

jiyeoon00 commented on src/main/java/com/project/coalba/domain/schedule/repository/ScheduleRepositoryImpl.java:
```diff
@@ -40,4 +41,61 @@ public List<Schedule> findAllByWorkspaceIdAndDateTimeRangeAndEndStatus(Long work
+    }
+
+    private LocalDateTime getStartTimeOf(LocalDate date) {
+        return date.atTime(0, 0, 0);
```
자정을 atTime(LocalTime.MIDNIGHT), 그 날의 끝시간을 atTime(LocalTime.MAX)로 나타내는 방법도 있는데 수동으로 치는것 보다 실수할 확률도 적고 더 가독성도 좋은 것 같아!

jkde7721 replied:
오 몰랐는데 고마워! 시간을 숫자 값으로 직접 입력해놓은 부분 다 고칠께

jkde7721 committed:
refactor: 스케줄 날짜시간 값 숫자 → 변수로 리팩토링

jkde7721 committed:
mod: schedule 최저시급 필드 기본값 변경

일단 직접 숫자값 입력
추후에 외부 API를 통해 값 받아오는 것으로 변경

jkde7721 committed:
refactor: 스케줄 관련 ServiceDto에서 의미 불분명한 Boolean 값이 아닌 Enum 값 가지도록 변경

HomeDateServiceDto, BossWorkspaceDateServiceDto에서
isSchedule, isAfterToday, isAllSuccess(Boolean 타입) ➡️  totalScheduleStatus(Enum 타입) 필드로 변경

jkde7721 committed:
refactor: schedule 도메인 service에서 @Transactional 메소드 레벨에 선언
